File: server/ai/mistral.py

import os
from mistralai.client import MistralClient
from mistralai.models.chat_completion import ChatMessage

class MistralAIService:
    def __init__(self):
        self.api_key = os.environ.get("MISTRAL_API_KEY")
        if not self.api_key:
            raise ValueError("MISTRAL_API_KEY environment variable is not set")
        self.client = MistralClient(api_key=self.api_key)
        
    def generate_text(self, prompt: str, model: str = "mistral-medium") -> str:
        """Generate text using Mistral AI models"""
        messages = [ChatMessage(role="user", content=prompt)]
        
        # Call the Mistral AI API
        chat_response = self.client.chat(
            model=model,
            messages=messages,
            temperature=0.7,
            max_tokens=1000
        )
        
        # Extract and return the generated text
        return chat_response.choices[0].message.content
    
    def generate_structured_data(self, prompt: str, model: str = "mistral-medium") -> dict:
        """Generate structured data response"""
        # Add instructions to format response as JSON
        formatted_prompt = f"{prompt} Please format your response as a JSON object."
        messages = [ChatMessage(role="user", content=formatted_prompt)]
        
        # Call the Mistral AI API
        chat_response = self.client.chat(
            model=model,
            messages=messages,
            temperature=0.7,
            max_tokens=1000
        )
        
        # Return the text response - will need to be parsed by caller
        return chat_response.choices[0].message.content
    
    def chat_conversation(self, messages: list[dict], model: str = "mistral-medium") -> str:
        """Have a multi-message conversation with Mistral AI"""
        # Convert dictionary messages to ChatMessage objects
        chat_messages = [
            ChatMessage(role=msg["role"], content=msg["content"]) 
            for msg in messages
        ]
        
        # Call the Mistral AI API
        chat_response = self.client.chat(
            model=model,
            messages=chat_messages,
            temperature=0.7,
            max_tokens=1000
        )
        
        # Return the generated response
        return chat_response.choices[0].message.content

# Example usage
if __name__ == "__main__":
    try:
        mistral_service = MistralAIService()
        response = mistral_service.generate_text("Explain the concept of color theory in design")
        print(response)
    except ValueError as e:
        print(f"Error: {e}")
        print("Please set the MISTRAL_API_KEY environment variable with your Mistral AI API key")

File: server/ai/mistral.ts

import { MistralClient } from '@mistralai/mistralai';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config();

// Initialize the Mistral AI client with API key from environment
const mistralClient = new MistralClient(process.env.MISTRAL_API_KEY || '');

/**
 * Generate text using Mistral AI
 */
export async function generateText(prompt: string, options = {}) {
  try {
    const defaultOptions = {
      model: 'mistral-large-latest',
      temperature: 0.7,
      maxTokens: 1000
    };

    const mergedOptions = { ...defaultOptions, ...options };

    const chatResponse = await mistralClient.chat({
      model: mergedOptions.model,
      messages: [{ role: 'user', content: prompt }],
      temperature: mergedOptions.temperature,
      maxTokens: mergedOptions.maxTokens
    });

    return chatResponse.choices[0]?.message?.content || '';
  } catch (error) {
    console.error('Error generating text with Mistral AI:', error);
    throw new Error(`Failed to generate text: ${(error as Error).message}`);
  }
}

/**
 * Generate structured JSON data with Mistral AI
 */
export async function generateJsonResponse<T>(prompt: string, systemPrompt = '', options = {}) {
  try {
    const defaultOptions = {
      model: 'mistral-large-latest',
      temperature: 0.3,
      maxTokens: 1500
    };

    const mergedOptions = { ...defaultOptions, ...options };

    // Create system message that enforces JSON output
    const finalSystemPrompt = 
      `${systemPrompt}\nYou must respond with valid JSON only, no other text.`;

    const chatResponse = await mistralClient.chat({
      model: mergedOptions.model,
      messages: [
        { role: 'system', content: finalSystemPrompt },
        { role: 'user', content: prompt }
      ],
      temperature: mergedOptions.temperature,
      maxTokens: mergedOptions.maxTokens
    });

    const content = chatResponse.choices[0]?.message?.content || '{}';
    
    // Parse JSON response
    return JSON.parse(content) as T;
  } catch (error) {
    console.error('Error generating JSON from Mistral AI:', error);
    throw new Error('Failed to generate valid JSON response');
  }
}

/**
 * Analyze content and provide insights
 */
export async function analyzeContent(content: string) {
  const prompt = `
    Analyze the following content and provide an assessment of:
    1. The overall sentiment (positive, negative, neutral)
    2. Key topics covered
    3. Important keywords for categorization

    Content: ${content}
  `;

  const systemPrompt = `
    You are an AI content analyst. Provide a JSON response with these fields:
    - sentiment: string (positive, negative, or neutral)
    - topics: string[] (array of main topics)
    - keywords: string[] (array of relevant keywords)
  `;

  return generateJsonResponse(prompt, systemPrompt);
}

export default {
  generateText,
  generateJsonResponse,
  analyzeContent,
  client: mistralClient
};
